pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+        .. code-block:: python3
+
+            measurements = [
+                qml.expval(qml.X(0) + qml.Hadamard(1)),
+                qml.expval(qml.X(0) + 0.2 * qml.Hadamard(1)),
+                qml.var(qml.Y(2) + qml.X(0)),
+            ]
+            tape = qml.tape.QuantumScript(measurements=measurements)
+            tapes, processing_fn = diagonalize_measurements(tape,
+                                                                 supported_base_obs=['PauliX', 'PauliY', 'PauliZ'])
+
+        Now ``tapes`` is a tuple containing a single tape with the updated measurements,
+        where only the Hadamard gate has been diagonalized:
+
+        >>> tapes[0].measurements
+        [expval(X(0) + Z(1)), expval(X(0) + 0.2 * Z(1)), var(Y(2) + X(0))]
+    """
+
+    bad_obs_input = [
+        o
+        for o in supported_base_obs
+        if o not in {"PauliX", "PauliY", "PauliZ", "Hadamard", "Identity"}
+    ]
+
+    if bad_obs_input:
+        raise ValueError(
+            "Supported base observables must be a subset of ['PauliX', 'PauliY', 'PauliZ', 'Hadamard'] "
+            f"but received {list(bad_obs_input)}"
+        )
+
+    supported_base_obs = set(list(supported_base_obs) + ["PauliZ", "Identity"])
+
+    _visited_obs = ([], [])  # tracks which observables and wires have been diagonalized
+    diagonalizing_gates = []
+    new_measurements = []
+
+    for m in tape.measurements:
+        if m.obs:
+            gates, new_obs, _visited_obs = _diagonalize_observable(
+                m.obs, _visited_obs, supported_base_obs
+            )
+            diagonalizing_gates.extend(gates)
+
+            meas = copy(m)
+            meas.obs = new_obs
+            new_measurements.append(meas)
+        else:
+            new_measurements.append(m)
+
+    new_operations = tape.operations + diagonalizing_gates
+
+    new_tape = type(tape)(
+        ops=new_operations,
+        measurements=new_measurements,
+        shots=tape.shots,
+        trainable_params=tape.trainable_params,
+    )
+
+    return (new_tape,), null_postprocessing
+
+
+def _check_if_diagonalizing(obs, _visited_obs, switch_basis):
+    """Checks if the observable should be diagonalized based on whether its basis should
+    be switched, and whether the same observable has already been diagonalized.
+
+    Args:
+        obs: the observable to be diagonalized
+        _visited_obs: a tuple containing a list of all observables that have already
+            been encountered, and a list of all wires that have already been encountered
+        switch_basis: whether the observable should be switched for one measuring in
+            the Z-basis on the same wires.
+
+    Returns:
+        diagonalize (bool): whether or not to apply diagonalizing gates for the observable
+        _visited_obs (tuple(Iterables): an up-to-date record of the observables and wires
+            encountered on the tape so far
+    """
+
+    if obs in _visited_obs[0] or isinstance(obs, qml.Identity):
```

Comment:
If we are primarily using `_visited_obs[0]` and `_visited_obs[1]` for checking whether something exists already, it might be worth using a set instead. Sets are optimized for quick `in` checks.

Reply:
changed it to a sets instead of lists